pkg/callerinfo/callerinfo.go

```diff
@@ -0,0 +1,42 @@
+// Copyright 2023 Outreach Corporation. All Rights Reserved.
+
+// Description: Provides the ability to discover the module+function who is calling you.
+
+// Package callerinfo provides the GetCallerFunction to get the name of the module and function
+// that has called you.
+package callerinfo
+
+import (
+	"fmt"
+	"runtime"
+)
+
+var moduleLookupByPC = make(map[uintptr]string)
+
+// Returns the name of the function, including module if applicable, that called GetCallerFunction.
+// skipFrames determines how many frames above that to skip.  If you want to know who is calling your
+// function, you would pass in skipFrames of 1.  skipFrames of 0 will return yourself.
+func GetCallerFunction(skipFrames uint16) (string, error) {
+	// We only care about the one stack frame above the caller, so skip at least two:
+	// 1. runtime.Callers
+	// 2. GetCallerModule
+	pc := make([]uintptr, 1)
+	skipTotal := 2 + int(skipFrames)
+	num := runtime.Callers(skipTotal, pc)
+	if num == 0 {
+		return "", fmt.Errorf("no frames returned from skip %d", skipTotal)
+	}
+
+	if mod, valid := moduleLookupByPC[pc[0]]; valid {
+		// Found it in the cache
+		return mod, nil
+	}
+
+	// Not cached -- have to do the slow lookup
+	frames := runtime.CallersFrames(pc)
+	frame, _ := frames.Next()
+
+	// Cache for later
+	moduleLookupByPC[pc[0]] = frame.Function
```

Comment:
You need a lock around the write here.

Reply:
Should we consider having this be an LRU cache?

Reply:
@jkinkead oh goddamnit, I thought it was both read and write concurrency safe, but it's only read.  Sigh.  Thanks.

@jaredallard It's capped to a count of all of the locations in the app making a log send, so I didn't feel like it made sense to add all the overhead of an LRU cache here.  Even if you have an outstandingly loud app with thousands of log locations, it's still going to be, worst case, like 100-200k of memory.

Reply:
Makes sense, just wanted to double-check that. I figured I was trying to overoptimize here :D